README.md
# TYPO3 Extension ``md_news_author``

This extension is based on extbase & fluid and provides the famous extension ``ext:news`` of Georg Ringer 
(thanks a lot @georgringer !) with one or more authors. You can centrally manage authors and attach them to 
news records. The extensions comes with two plugins, one for listing all authors and one for the detail page of an 
author which also shows the news records of the selected author.

## Requirements

- TYPO3 >= 10.4
- ext:news >= 7.0
- ext:numbered_pagination >= 1.0

### Installation

- Install the extension by using the extension manager or use composer (`composer req mediadreams/md_news_author`)
- Include the static TypoScript of the extension
- Configure the extension by setting your own constants

## Usage

### Create authors and attach them to news records

- Create some author records on a sysfolder (use list modul, push plus-icon `Create new record` and select `News Author`)
- Create a news record on a sysfolder and find the new tab `Author`
- Select one or more authors for the news record
- Save and close

### List authors

Insert paginated list of all authors.

- Create a plugin `News author: Author list` on a page
- Choose for `Page with single author view` the page with single author view
- Select the sysfolder where the author records are stored
- Additional settings can be found in the tab `List view settings`
- If needed, show authors of certain categories only (tab `Categories`)
- Save and close

### Authors detail page

Insert an author detail view. This page includes also all news which are associated with the choosen author.

- Create a plugin `News author: Show author` on a page
- Optionally choose for `Page with author list` the page with the list of all authors
- Select the sysfolder where the author records are stored
- Additional settings can be found in the tab `Detail view settings`
- Save and close

### Show author in ``ext:news`` view

- Access the author properties in a news record with `{newsItem.newsAuthor}`. Since there could be more 
than one author attached to a news record, you have to iterate:

```
<f:for each="{newsItem.newsAuthor}" as="author">
    {md:ShowAuthorName(author: author)}
    {author.phone}
    {author. ...}
</f:for>
```

- Add a link to the profile page

Don't forget to load the viewhelper `{namespace md=Mediadreams\MdNewsAuthor\ViewHelpers}`:

    <f:for each="{newsItem.newsAuthor}" as="author">
        <f:link.action action="show" controller="NewsAuthor" extensionName="mdnewsauthor" pluginName="show" arguments="{newsAuthor: author}" pageUid="{settings.newsAuthor.authorDetailPid}" title="More about {md:ShowAuthorName(author:'{author}')}">
            <md:ShowAuthorName author="{author}" />
        </f:link.action>
    </f:for>

### Page TSconfig

In order to show only authors of a single page in the "Authors"-tab of a news record, you can use the following TSconfig:

    TCEFORM.tx_news_domain_model_news.news_author.PAGE_TSCONFIG_STR = 1

This will show only the author records, which are stored on page ID = 1

### ``routeEnhancers``

```
routeEnhancers:
  NewsAuthorList:
    type: Extbase
    extension: MdNewsAuthor
    plugin: list
    routes:
      -
        routePath: '/a-z/{letter}'
        _controller: 'NewsAuthor::list'
        _arguments:
          'letter': 'selectedLetter'
    defaultController: 'NewsAuthor::list'
  NewsAuthorShow:
    type: Extbase
    extension: MdNewsAuthor
    plugin: show
    routes:
      - 
        routePath: '{slug}'
        _controller: 'NewsAuthor::show'
        _arguments:
          'slug': 'newsAuthor'
    requirements:
      slug: '^[a-zA-Z0-9].*$'
    aspects:
      slug:
        type: PersistedAliasMapper
        tableName: 'tx_mdnewsauthor_domain_model_newsauthor'
        routeFieldName: 'slug'
      letter:
        type: StaticValueMapper
        map:
          a: A
          b: B
          c: C
          d: D
          e: E
          f: F
          g: G
          h: H
          i: I
          j: J
          k: K
          l: L
          m: M
          n: N
          o: O
          p: P
          q: Q
          r: R
          s: S
          t: T
          u: U
          v: V
          w: W
          x: X
          y: Y
          z: Z
```

## Bugs and Known Issues
If you find a bug, it would be nice if you add an issue on [Github](https://github.com/cdaecke/md_news_author/issues).

# THANKS

Thanks a lot to all who make this outstanding TYPO3 project possible!

## Credits

Extension icon by [Font Awesome](https://fontawesome.com/icons/user?style=solid).


Changelog
Version 6.0.1 (2021-10-06)
    TASK: Update TYPO3 dependency in composer.json

    All changes
    https://github.com/cdaecke/md_news_author/compare/v6.0.0...v6.0.1

Version 6.0.0 (2021-10-05)
    FEATURE: TYPO3 11 compatibility
    TASK: Dependency to ext:numbered_pagination was added.

    BREAKING
    - News Author-Plugin was splited in two separate plugins, one for listing the authors and one for a detail view of an author. Use the Upgrade Wizard to migrate!
    - Templates where changed.
    - ShowAlphabeticalNavigationViewHelper was removed.
    - New paginator was introduced, so partial needs to be change, if you use your own.
    - routeEnhancers for speaking urls need to be adapted (find new configuration in the documentation).

    All changes
    https://github.com/cdaecke/md_news_author/compare/v5.0.1...v6.0.0

Version 5.0.0 (2020-09-16)
    [BUGFIX] get author data in news record again (with ext:news v8.4.0)

    All changes
    https://github.com/cdaecke/md_news_author/compare/v5.0.0...v5.0.1

Version 5.0.0 (2020-08-04)
    FEATURE: TYPO3 10 compatibility

    All changes
    https://github.com/cdaecke/md_news_author/compare/v4.0.1...v5.0.0

Version 4.0.1 (2020-01-07)
    BUGFIX: Get all news records on an authors detail page

    All changes
    https://github.com/cdaecke/md_news_author/compare/v4.0.0...v4.0.1

Version 4.0.0 (2019-02-14)
    FEATURE: TYPO3 9 compatibility

    ATTENTION:
    Make sure to run the database analyzer in the install tool and clear the cache afterwards!

    All changes
    https://github.com/cdaecke/md_news_author/compare/v3.0.2...v4.0.0

Version 3.0.2 (2019-02-04)
    BUGFIX: Respect translations of authors in news records
    BUGFIX: Make TCEFORM working again for TYPO3 8
    Add some more fields the author record

    ATTENTION: Make sure to clear the cache in TYPO3 install tool after this update!

    All changes
    https://github.com/cdaecke/md_news_author/compare/v3.0.1...v3.0.2

Version 3.0.1 (2018-02-19)
    BUGFIX: Do not allow to add the same author multiple times to the same news record

    All changes
    https://github.com/cdaecke/md_news_author/compare/v3.0.0...v3.0.1

Version 3.0.0 (2017-11-30)
    Add possibility to have more than one author attached to a news record

    All changes
    https://github.com/cdaecke/md_news_author/compare/v2.0.2...v3.0.0

Version 2.0.2 (2017-07-10)
    Fix error in list view

    All changes
    - 2017-07-10 [TASK] raise version number (Commit 8e65802 by chris)
    - 2017-07-10 [BUGFIX] initialize arguments in ShowAlphabeticalNavigationViewHelper (Commit 1886270 by chris)

Version 2.0.1 (2017-07-10)
    - add composer.json
    - fix ShowAlphabeticalNavigationViewHelper()

    All changes
    - 2017-07-10 [TASK] raise version number (Commit 083177d by chris)
    - 2017-07-10 [BUGFIX] remove PHP warning in ShowAlphabeticalNavigationViewHelper(). Make it compatible with AbstractConditionViewHelper::render() (Commit b0eced4 by chris)
    - 2017-06-28 [BUGFIX] composer.json (Commit 9d6d438 by chris)
    - 2017-06-28 [DOC] add section "Bugs and Known Issues" (Commit f8599d1 by chris)
    - 2017-06-28 [TASK] add composer.json (Commit b76a1f2 by chris)


Version 2.0.0 (2017-06-14)
    - add alphabetical filter for author list
    - add categories for authors

    All changes:
    - 2017-06-14 [DOC] add categories to documentation (Commit f801e27 by chris)
    - 2017-06-14 [TASK] update version number (Commit 7c47626 by chris)
    - 2017-06-05 [FEATURE] add alphabetical filter to author list (Commit af24452 by chris)
    - 2017-06-04 [TASK] update documentation: add list view to realURL config (Commit f75a9eb by chris)
    - 2017-05-14 [TASK] add tags for ext:indexed_search in templates (<!--TYPO3SEARCH_end--> and <!--TYPO3SEARCH_begin-->) - Do not index the author list - Do not index the articles on authors detail page (Commit ecb4ec3 by chris)
    - 2017-05-12 [TASK] use own pagination template (copy of the news pagination - thanks Georg Ringer!) (Commit 0fbe573 by chris)
    - 2017-05-08 [TASK] add categories to authors, filter authors by categories references #1 (Commit f50f2aa by chris)
    - 2017-05-06 [TASK] add pagination widget to list view (Commit 482f5d1 by chris)
    - 2017-05-06 [TASK] change version number to DEV (Commit c54a599 by chris)


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext: "md_news_author"
 *
 * Auto generated by Extension Builder 2016-09-15
 *
 * Manual updates:
 * Only the data in the array - anything else is removed by next write.
 * "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = [
    'title' => 'News Author',
    'description' => 'Adds one or more authors to a tx_news record, show a list of all authors and display a detail page of the author containing the attached news entries.',
    'category' => 'plugin',
    'author' => 'Christoph Daecke',
    'author_email' => 'typo3@mediadreams.org',
    'state' => 'stable',
    'internal' => '',
    'uploadfolder' => '0',
    'createDirs' => '',
    'clearCacheOnLoad' => 0,
    'version' => '6.0.3',
    'constraints' => [
        'depends' => [
            'typo3' => '10.4.0-11.5.99',
            'news' => '7.0-9.99',
            'numbered_pagination' => '1.0-1.99',
        ],
        'conflicts' => [],
        'suggests' => [],
    ],
];
